pipeline/metadata/flatten_satellite.py

```diff
@@ -75,111 +76,146 @@ def format_timestamp(timestamp: str) -> str:
     scan: a loaded json object containing the parsed content of the line
+    filepath: path like "<path>/<filename>.json.gz"
 
   Yields:
     Rows
   """
-  responses = scan.get('response', [])
-  row['controls_failed'] = not scan['passed_liveness']
-  row['rcode'] = [str(response['rcode']) for response in responses]
-
-  row['average_confidence'] = scan.get('confidence')['average']
-  row['matches_confidence'] = scan.get('confidence')['matches']
-  row['untagged_controls'] = scan.get('confidence')['untagged_controls']
-  row['untagged_response'] = scan.get('confidence')['untagged_response']
-
-  row['excluded'] = scan.get('excluded', False)
-  row['exclude_reason'] = ' '.join(scan.get('exclude_reason', []))
-
-  errors = [
-      response['error']
-      for response in responses
-      if response['error'] and response['error'] != 'null'
-  ]
-  row['error'] = ' | '.join(errors) if errors else None
-  yielded = False
-  has_test_domain = False
-  for response in responses:
-    if response['url'] == row['domain']:
-      has_test_domain = True
-    # Check responses for test domain with valid answers
-    if response['url'] == row['domain'] and (response['rcode'] == 0 and
-                                             response['has_type_a']):
-      row['has_type_a'] = True
-      # Separate into one answer IP per row for tagging
-      row['received'] = []
-      for ip in response['response']:
-        received = {
-            'ip': ip,
-            'http': response['response'][ip].get('http'),
-            'cert': response['response'][ip].get('cert'),
-            'asname': response['response'][ip].get('asname'),
-            'asnum': response['response'][ip].get('asnum'),
-            'matches_control': ''
-        }
-        matched = response['response'][ip].get('matched', [])
-        if matched:
-          received['matches_control'] = ' '.join(
-              [tag for tag in matched if tag in SATELLITE_TAGS])
-        row['received'].append(received)
-      yielded = True
-      yield row.copy()
-    # Do not output rows if there are no trials for the test domain.
-  if not yielded and has_test_domain:
-    # All trials for the test domain in `scan` are errors
-    yield row.copy()
+  row['controls_failed'] = not scan['passed_control']
+  received_ips = scan.get('response').copy()
+
+  if not received_ips:
+    row['received'] = []
+    yield row
+    return
+
+  # separate into one answer ip per row for tagging
+  rcodes = received_ips.pop('rcode', [])
+  errors = received_ips.pop('error', [])
+
+  successful_test_rcode, failed_test_rcodes = split_rcodes(rcodes)
+
+  if any(rcode == "0" for rcode in failed_test_rcodes):
+    raise Exception(
+        f"Satellite v2.1 measurement has multiple 0 rcodes: {filepath} - {scan}"
+    )
+
+  if not successful_test_rcode and received_ips:
+    # TODO figure out how to handle this bug case from 2021-05-16
+    # for now drop the measurements
+    logging.warning(
```

Comment:
Perhaps introduce an error counter

Reply:
We can count these logs in a job that succeeded. There are 4070 rows that trigger this error in the whole dataset.

![6fGL4r5iN8nsVjf](https://user-images.githubusercontent.com/1127209/152400925-87cda0ae-5b40-48bc-a94a-767f2ccba360.png)

https://console.cloud.google.com/logs/query;query=resource.type%3D%22dataflow_step%22%0Aresource.labels.job_id%3D%222022-01-17_20_58_59-987126952126004496%22%0Aseverity%3DWARNING%0A%22Satellite%20v2.1%20measurement%20has%20ips%20but%20no%200%20rcode%22;timeRange=2022-01-18T04:58:59.933Z%2F2022-01-18T13:01:53.036Z;cursorTimestamp=2022-01-18T11:06:33.445386171Z?project=firehook-censoredplanet

We don't expect this number to ever get any larger (since no more 2.1 data is being generated.) So this number will never get bigger. Implementing a counter in code doesn't seem like a good use of time.